distributed/comm/mpi.py

```diff
@@ -0,0 +1,781 @@
+                header = struct.unpack(header_fmt, header)
+                cuda_frames, sizes = header[:nframes], header[nframes:]
+            except (Exception, CancelledError):
+                self.abort()
+                raise CommClosedError("While reading, the connection was closed")
+            else:
+                # Recv frames
+                frames = [
+                    device_array(each_size) if is_cuda else host_array(each_size)
+                    for is_cuda, each_size in zip(cuda_frames, sizes)
+                ]
+                cuda_recv_frames, recv_frames = zip(
+                    *(
+                        (is_cuda, each_frame)
+                        for is_cuda, each_frame in zip(cuda_frames, frames)
+                        if len(each_frame) > 0
+                    )
+                )
+
+                # It is necessary to first populate `frames` with CUDA arrays 
+                # and synchronize the default stream before starting receiving 
+                # to ensure buffers have been allocated
+                if any(cuda_recv_frames):
+                    synchronize_stream(0)
+
+
+                sizes_new = tuple(i for i in sizes if i != 0) # remove all 0's from this list
+                assert len(sizes_new) < TAG_QUOTA_PER_CONNECTION - 2
+
+                logger.debug("read_3: me=%s, you=%s, tag=%s, sizes=%s, sizes_new=%s, rf_len=%s, crf_len=%s", \
+                                         MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag, sizes, sizes_new, len(recv_frames), len(cuda_recv_frames))
+
+                read_counter = 0
+                for each_frame in recv_frames:
+                    #await self.ep.recv(each_frame)
+                    await self.mpi_recv(each_frame, sizes_new[read_counter], tag)
+                    tag = tag + 1
+                    read_counter = read_counter + 1
+
+                logger.debug("read_4: me=%s, you=%s, tag=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag)
+
+                msg = await from_frames(
+                    frames,
+                    deserialize=self.deserialize,
+                    deserializers=deserializers,
+                    allow_offload=self.allow_offload,
+                )
+                logger.debug("read_5: me=%s, you=%s, tag=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag)
+                return msg
+
+    async def close(self):
+        if self._closed is False:
+            self._closed = True
+
+    def abort(self):
+        if self._closed is False:
+            self._closed = True
+
+    def closed(self):
+        return self._closed
+
+
+class MPI4DaskConnector(Connector):
+    prefix = "mpi://"
+    comm_class = MPI4Dask
+    encrypted = False
+
+    async def connect(self, address: str, \
+                      deserialize=True, \
+                      **connection_args) -> MPI4Dask:
+
+        rand_num = 0.001 * random.randint(0, 1000)
+        await asyncio.sleep(rand_num)
```

Comment:
Why is a random sleep needed?

Reply:
I explained this earlier. We should be able to avoid this sleep. I will update here on this.